tests/kernel/interrupt/src/interrupt_offload.c

```diff
@@ -0,0 +1,266 @@
+	}
+}
+
+static void t_running(void *p1, void *p2, void *p3)
+{
+	k_sem_give(&start_sem);
+
+	while (1) {
+		orig_t_keep_run = 1;
+		k_usleep(1);
+	}
+}
+
+static void run_test_offload(int case_type, int real_irq)
+{
+	static bool wq_already_start;
+	int thread_prio = 1;
+
+	printk("case %d\n", case_type);
+
+	/* two semaphore used to sync start and end */
+	k_sem_init(&start_sem, 0, 1);
+	k_sem_init(&end_sem, 0, NUM_WORK);
+
+	if (offload_job_prio_higher) {
+		thread_prio = 2;
+	}
+
+	if (real_irq) {
+		init_dyn_interrupt();
+	}
+
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+			(k_thread_entry_t)t_running,
+			NULL, NULL, NULL,
+			K_PRIO_PREEMPT(thread_prio),
+			K_INHERIT_PERMS, K_NO_WAIT);
+
+	/* start a work queue thread if not existing */
+	if (!wq_already_start) {
+		k_work_queue_start(&wq_queue, wq_stack, STACK_SIZE,
+		1, NULL);
+
+		wq_already_start = true;
+	}
+
+	/* wait for thread start */
+	k_sem_take(&start_sem, K_FOREVER);
+
+	for (int i = 0; i < NUM_WORK; i++) {
+
+		switch (case_type) {
+		case TEST_OFFLOAD_MULTI_JOBS:
+			trigger_offload_interrupt(real_irq,
+					(void *)&offload_work[i]);
+		break;
+		case TEST_OFFLOAD_INDETICAL_JOBS:
+			trigger_offload_interrupt(real_irq,
+					(void *)&offload_work[0]);
+
+			/*
+			 * An 1 us delay for offload the same job is
+			 * needed, or it will cause an infinite loop.
+			 */
+			k_usleep(1);
+		break;
+		default:
+			ztest_test_fail();
+		}
+	}
+
+	/* wait for all offload job complete */
+	k_sem_take(&end_sem, K_FOREVER);
+
+	k_usleep(1);
+
+	zassert_equal(orig_t_keep_run, 1,
+			"offload job done, the original thread run");
+
+	k_thread_abort(tid);
+}
+
+/**
+ * @brief Test interrupt offload work to multiple jobs
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details Validate isr can offload workload to multi work queue, and:
+ *
+ * - If the priority of the original thread < offload job, offload jobs
+ *   could execute immediately.
+ *
+ * - If the priority of the original thread > offload job, offload jobs
+ *   will not execute immediately.
+ *
+ * We test this by irq_offload().
+ */
+void test_isr_offload_job_multiple(void)
+{
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_MULTI_JOBS, false);
+
+	offload_job_prio_higher = false;
+	run_test_offload(TEST_OFFLOAD_MULTI_JOBS, false);
+}
+
+/**
+ * @brief Test interrupt offload work to identical jobs
+ *
+ * @ingroup kernel_interrupt_tests
+ *
+ * @details Validate isr can offload workload to work queue, and all
+ * the offload jobs use the same thread entry, and:
+ *
+ * - If the priority of the original thread < offload job, offload jobs
+ *   could execute immediately.
+ *
+ * - If the priority of the original thread > offload job, offload jobs
+ *   will not execute immediately.
+ *
+ * We test this by irq_offload().
+ */
+void test_isr_offload_job_identi(void)
+{
+	offload_job_prio_higher = true;
+	run_test_offload(TEST_OFFLOAD_INDETICAL_JOBS, false);
```

Comment:
Here too, IDENTICAL.

Reply:
Thanks for correcting this. Got it fixed.